pkg/controller/device.go

```diff
@@ -367,54 +375,76 @@ func (cloud *CloudCtx) checkDrive(drive *config.Drive, dataStores []*config.Data
 	return dataStores, nil
 }
 
-func imageToContentTree(image *config.Image) *config.ContentTree {
-	return &config.ContentTree{
-		Uuid:            image.Uuidandversion.Uuid,
+func (cloud *CloudCtx) imageToContentTree(image *config.Image, displayName string) *config.ContentTree {
+	for _, el := range cloud.ListContentTree() {
+		if el.DisplayName == displayName && el.URL == image.Name && el.Sha256 == image.Sha256 && el.DsId == image.DsId {
+			return el
+		}
+	}
+	id, err := uuid.NewV4()
+	if err != nil {
+		log.Fatal(err)
+	}
+	contentTree := &config.ContentTree{
+		Uuid:            id.String(),
 		DsId:            image.DsId,
 		URL:             image.Name,
 		Iformat:         image.Iformat,
 		Sha256:          image.Sha256,
 		MaxSizeBytes:    uint64(image.SizeBytes),
 		Siginfo:         image.Siginfo,
-		DisplayName:     image.Name,
+		DisplayName:     displayName,
 		GenerationCount: 0,
 	}
+	_ = cloud.AddContentTree(contentTree)
+	return contentTree
 }
 
-func checkContentTree(contentTree *config.ContentTree, contentTrees []*config.ContentTree) []*config.ContentTree {
+func checkContentTree(contentTree *config.ContentTree, contentTrees []*config.ContentTree) (*config.ContentTree, []*config.ContentTree) {
 	for _, el := range contentTrees {
 		if el.Uuid == contentTree.Uuid {
-			return contentTrees
+			return el, contentTrees
 		}
 	}
-	return append(contentTrees, contentTree)
+	return contentTree, append(contentTrees, contentTree)
 }
 
-func driveToVolume(dr *config.Drive) *config.Volume {
-	return &config.Volume{
-		Uuid: dr.Image.Uuidandversion.Uuid,
+func (cloud *CloudCtx) driveToVolume(dr *config.Drive, contentTree *config.ContentTree) *config.Volume {
+	for _, el := range cloud.ListVolume() {
+		if el.Origin.DownloadContentTreeID == contentTree.Uuid && el.DisplayName == fmt.Sprintf("%s_0_m_0", contentTree.DisplayName) {
+			return el
+		}
+	}
+	id, err := uuid.NewV4()
+	if err != nil {
+		log.Fatal(err)
+	}
+	volume := &config.Volume{
+		Uuid: id.String(),
 		Origin: &config.VolumeContentOrigin{
 			Type:                  config.VolumeContentOriginType_VCOT_DOWNLOAD,
-			DownloadContentTreeID: dr.Image.Uuidandversion.Uuid,
+			DownloadContentTreeID: contentTree.Uuid,
 		},
 		Protocols:    nil,
 		Maxsizebytes: dr.Maxsizebytes,
-		DisplayName:  dr.Image.Name,
+		DisplayName:  fmt.Sprintf("%s_0_m_0", contentTree.DisplayName),
 	}
+	_ = cloud.AddVolume(volume)
+	return volume
 }
 
-func checkVolume(vol *config.Volume, vols []*config.Volume) []*config.Volume {
+func checkVolume(vol *config.Volume, vols []*config.Volume) (*config.Volume, []*config.Volume) {
```

Comment:
same question.. vol *config.Volume, vols []*config.Volume) (*config.Volume, []*config.Volume) { and no comment.!!!